src/hash_traits.jl

```diff
@@ -0,0 +1,461 @@
+    else
+        return qualified_name_(T)
+    end
+end
+
+#####
+##### DataType
+#####
+
+sorted_field_names(T::Type) = TupleTools.sort(fieldnames(T); by=string)
+@generated function sorted_field_names(T)
+    return TupleTools.sort(fieldnames(T); by=string)
+end
+
+function type_structure(::Type{T}, trait::StructTypes.DataType) where {T}
+    if isconcretetype(T)
+        fields = trait isa StructTypes.OrderedStruct ? fieldnames(T) : sorted_field_names(T)
+        return fields, map(field -> fieldtype(T, field), fields)
+    else
+        return nothing
+    end
+end
+
+function stable_hash_helper(x, hash_state, context, st::StructTypes.DataType)
+    nested_hash_state = start_nested_hash!(hash_state)
+
+    # hash the field values
+    fields = st isa StructTypes.UnorderedStruct ? sorted_field_names(x) :
+             fieldnames(typeof(x))
+    nested_hash_state = hash_fields(x, fields, nested_hash_state, context)
+    hash_state = end_nested_hash!(hash_state, nested_hash_state)
+    return hash_state
+end
+
+function hash_fields(x, fields, hash_state, context)
+    for field in fields
+        val = getfield(x, field)
+        # can we optimize away the field's type_hash?
+        transform = transformer(typeof(val), context)
+        if isconcretetype(fieldtype(typeof(x), field)) && transform.preserves_structure
+            # the fieldtype has been hashed as part of the type of the container
+            hash_value(val, hash_state, context, transform)
+        else
+            hash_type_and_value(val, hash_state, context)
+        end
+    end
+    return hash_state
+end
+
+#####
+##### ArrayType
+#####
+
+is_ordered(x) = true
+is_ordered(::AbstractSet) = false
+order_by(x::Symbol) = String(x)
+order_by(x::Char) = string(x)
+order_by(x) = x
+
+function type_structure(::Type{T}, ::StructTypes.ArrayType) where {T}
+    return eltype(T)
+end
+
+# include ndims in type hash when we can
+function type_structure(::Type{T}, ::StructTypes.ArrayType) where {T<:AbstractArray}
+    return eltype(T), ndims_(T)
+end
+ndims_(::Type{<:AbstractArray{<:Any,N}}) where {N} = N
+ndims_(::Type{<:AbstractArray}) = nothing
+
+function transformer(::Type{<:AbstractRange}, ::HashVersion{3})
+    return Transformer(identity, StructTypes.Struct(); preserves_structure=true)
+end
+
+function transformer(::Type{<:AbstractArray}, ::HashVersion{3})
+    return Transformer(x -> (size(x), split_union(x)); preserves_structure=true)
+end
+
+split_union(array) = TransformIdentity(array)
+# NOTE: this method actually properly handles union splitting for as many splits as julia
+# will allow to match to this method, not just two; in the case where the eltype is
+# Union{Int, UInt, Char} for instance, M will match to Union{UInt, Char} and the `else`
+# branch will properly split out the first type. The returned M_array will then be split
+# again, when the `transformer` method above is applied to it.
+function split_union(array::AbstractArray{Union{N,M}}) where {N,M}
```

Comment:
btw do we get different hashes for `Vector{Union{N,M}}` vs `Vector{Union{M,N}}`?

If so that's mabye OK but we should document it

Reply:
As best I can tell, the current julia implementation of `Union` types systematically re-orders the elements of the union, such that user specified ordering is irrelevant.

```julia
julia> using StableHashTraits

julia> struct A
       val::Int
       end

julia> struct B
       val::Float64
       end

julia> Union{A,B}[A(1), B(2)]
2-element Vector{Union{A, B}}:
 A(1)
 B(2.0)

julia> Union{B,A}[A(1), B(2)]
2-element Vector{Union{A, B}}:
 A(1)
 B(2.0)

julia> Union{B,A}[B(2), A(1)]
2-element Vector{Union{A, B}}:
 B(2.0)
 A(1)

julia> bytes2hex(stable_hash(Union{A,B}[A(1), B(2)]))
"c122260e70fb39dbd4d0340de5dc839886444352b85026c0e99fcfecb5c3ee56"

julia> bytes2hex(stable_hash(Union{B,A}[A(1), B(2)]))
"c122260e70fb39dbd4d0340de5dc839886444352b85026c0e99fcfecb5c3ee56"

julia> Union{A,B} === Union{B,A}
true
```